/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2023-11-23T19:46:56.418011
 */
package bio.ferlab.clin.etl.model.raw

case class RawFranklin(`annotations`: ANNOTATIONS = ANNOTATIONS(),
                       `classification`: CLASSIFICATION = CLASSIFICATION(),
                       `included_in_analysis_workbench`: Boolean = false,
                       `phenotypes`: Seq[String] = Seq("global developmental delay", "intellectual disability", "coarse facial features", "delayed speech and language development", "abnormality of the dentition", "autistic behavior", "strabismus"),
                       `priority`: PRIORITY = PRIORITY(),
                       `sample_data`: SAMPLE_DATA = SAMPLE_DATA(),
                       `variant`: VARIANT = VARIANT(),
                       `variant_franklin_link`: String = "https://franklin.genoox.com/clinical-db/variant/snp/chr16-2498261-AT-A-HG38",
                       `batch_id`: String = "1",
                       `family_id`: String = "1",
                       `aliquot_id`: String = "12345",
                       `analysis_id`: String = "20")

case class PREDICTIONS(`ada`: Double = 0.968532953361239,
                       `aggregated_predictions`: Double = 0.9879213483146067,
                       `dbsnp`: String = "rs61741107",
                       `fathmm`: Double = -4.59,
                       `fitcons`: Double = 0.580535,
                       `genocanyon`: Double = 0.999999999999996,
                       `gerp`: Double = 4.59,
                       `ma`: Double = 3.695,
                       `metalr`: Double = 0.9657,
                       `mt`: Double = 1.0,
                       `revel`: Double = 0.957,
                       `sift`: Double = 0.0,
                       `splice_ai`: Double = 0.0)

case class ANNOTATIONS(`clinical_evidences`: CLINICAL_EVIDENCES = CLINICAL_EVIDENCES(),
                       `frequencies`: FREQUENCIES = FREQUENCIES(),
                       `predictions`: PREDICTIONS = PREDICTIONS(),
                       `transcripts`: Seq[TRANSCRIPTS] = Seq(TRANSCRIPTS()))

case class TRANSCRIPTS(`closest_distance_to_exon`: Long = 0,
                       `closest_exon`: Long = 4,
                       `effect`: String = "FRAMESHIFT",
                       `exon_number`: Long = 4,
                       `gene`: String = "TBC1D24",
                       `hgvc_c`: String = "c.1008del",
                       `hgvc_p`: String = "p.His336GlnfsTer12",
                       `region`: String = "EXONIC",
                       `transcript`: String = "NM_001199107.2",
                       `transcript_exon_count`: Long = 8,
                       `transcript_type`: String = "REFSEQ",
                       `transcripts_count`: Long = 2)

case class CLINVAR(`submissions_by_classification`: Seq[SUBMISSIONS_BY_CLASSIFICATION] = Seq(SUBMISSIONS_BY_CLASSIFICATION()))

case class VARIANT(`alt`: String = "A",
                   `chromosome`: String = "chr16",
                   `end_position`: Long = 2498262,
                   `ref`: String = "AT",
                   `start_position`: Long = 2498261,
                   `variation_type`: String = "INDEL")

case class ACMG_RULES(`influence`: String = "StrongPathogenic",
                      `is_met`: Boolean = false,
                      `name`: String = "PS1",
                      `weight`: Long = 7)

case class CLINICAL_EVIDENCES(`clinvar`: CLINVAR = CLINVAR(),
                              `uniprot`: UNIPROT = UNIPROT())

case class SUBMISSIONS_BY_CLASSIFICATION(`count`: Long = 11,
                                         `level`: String = "PATHOGENIC")

case class UNIPROT(`uniprot_id`: String = "P15918",
                   `variant_link`: String = "https://www.uniprot.org/uniprot/P15918")

case class CLASSIFICATION(`acmg_classification`: String = "PATHOGENIC",
                          `acmg_rules`: Seq[ACMG_RULES] = Seq(ACMG_RULES()))

case class FAMILY_ZYGOSITIES(`father_zygosity`: String = "HOM_REF",
                             `mother_zygosity`: String = "HET")

case class FREQUENCIES(`aggregated_frequency`: Double = 5.274287E-5,
                       `max_gnomad_frequency`: Double = 1.1763667862396687E-4)

case class SAMPLE_DATA(`confidence_level`: String = "MEDIUM",
                       `depth`: Long = 101,
                       `depth_alt`: Long = 52,
                       `depth_ref`: Long = 49,
                       `family_inheritance_model`: String = "DE NOVO",
                       `family_zygosities`: FAMILY_ZYGOSITIES = FAMILY_ZYGOSITIES(),
                       `gq`: Double = 50.0,
                       `mapping_quality`: Double = 248.25,
                       `pl`: String = "85:0:50",
                       `quality`: Double = 76.29,
                       `vaf`: Double = 0.5148514851485149,
                       `zygosity`: String = "HET")

case class PRIORITY(`score`: Double = 0.9089091283319308)