package bio.ferlab.clin.model

case class CancerHotspotsOutput(`chromosome`: String = "1",
                                `start`: Long = 119032994,
                                `end`: Long = 119032995,
                                `hugo_symbol`: String = "WARS2",
                                `entrez_gene_id`: String = "10352",
                                `center`: String = ".",
                                `ncbi_build`: String = "GRCh37",
                                `strand`: String = "+",
                                `variant_classification`: String = "Missense_Mutation",
                                `variant_type`: String = "SNP",
                                `reference_allele`: String = "C",
                                `tumor_seq_allele1`: String = "C",
                                `tumor_seq_allele2`: String = "T",
                                `dbsnp_rs`: String = "novel",
                                `dbsnp_val_status`: Option[String] = None,
                                `tumor_sample_barcode`: String = "000236",
                                `matched_norm_sample_barcode`: String = "NORMAL",
                                `match_norm_seq_allele1`: String = "C",
                                `match_norm_seq_allele2`: String = "C",
                                `tumor_validation_allele1`: Option[String] = None,
                                `tumor_validation_allele2`: Option[String] = None,
                                `match_norm_validation_allele1`: Option[String] = None,
                                `match_norm_validation_allele2`: Option[String] = None,
                                `verification_status`: Option[String] = None,
                                `validation_status`: Option[String] = None,
                                `mutation_status`: Option[String] = None,
                                `sequencing_phase`: Option[String] = None,
                                `sequence_source`: Option[String] = None,
                                `validation_method`: Option[String] = None,
                                `score`: Option[String] = None,
                                `bam_file`: Option[String] = None,
                                `sequencer`: Option[String] = None,
                                `tumor_sample_uuid`: Option[String] = None,
                                `matched_norm_sample_uuid`: Option[String] = None,
                                `hgvsc`: String = "c.1000G>A",
                                `hgvsp`: String = "p.Val334Ile",
                                `hgvsp_short`: String = "p.V334I",
                                `transcript_id`: String = "ENST00000235521",
                                `exon_number`: String = "6/6",
                                `t_depth`: String = "0",
                                `t_ref_count`: String = ".",
                                `t_alt_count`: String = ".",
                                `n_depth`: String = "0",
                                `n_ref_count`: String = ".",
                                `n_alt_count`: String = ".",
                                `all_effects`: String = "WARS2,missense_variant,p.Val334Ile,ENST00000235521,NM_201263.2,NM_015836.3;WARS2,missense_variant,p.Val240Ile,ENST00000537870,;WARS2,3_prime_UTR_variant,,ENST00000369426,;WARS2,downstream_gene_variant,,ENST00000497402,;WARS2,downstream_gene_variant,,ENST00000495746,;",
                                `allele`: String = "T",
                                `gene`: String = "ENSG00000116874",
                                `feature`: String = "ENST00000235521",
                                `feature_type`: String = "Transcript",
                                `consequence`: String = "missense_variant",
                                `cdna_position`: String = "1027/2800",
                                `cds_position`: String = "1000/1083",
                                `protein_position`: String = "334/360",
                                `amino_acids`: String = "V/I",
                                `codons`: String = "Gtt/Att",
                                `existing_variation`: Option[String] = None,
                                `allele_num`: String = "1",
                                `distance`: Option[String] = None,
                                `strand_vep`: String = "-1",
                                `symbol`: String = "WARS2",
                                `symbol_source`: String = "HGNC",
                                `hgnc_id`: String = "12730",
                                `biotype`: String = "protein_coding",
                                `canonical`: String = "YES",
                                `ccds`: String = "CCDS900.1",
                                `ensp`: String = "ENSP00000235521",
                                `swissprot`: String = "Q9UGM6",
                                `trembl`: String = "B7Z5X7",
                                `uniparc`: String = "UPI000004A002",
                                `refseq`: String = "NM_201263.2,NM_015836.3",
                                `sift`: String = "tolerated(0.31)",
                                `polyphen`: String = "benign(0.015)",
                                `exon`: String = "6/6",
                                `intron`: Option[String] = None,
                                `domains`: String = "Gene3D:1.10.240.10,HAMAP:MF_00140_B,hmmpanther:PTHR10055,Low_complexity_(Seg):seg,Superfamily_domains:SSF52374,TIGRFAM_domain:TIGR00233",
                                `clin_sig`: Option[String] = None,
                                `somatic`: Option[String] = None,
                                `pubmed`: Option[String] = None,
                                `motif_name`: Option[String] = None,
                                `motif_pos`: Option[String] = None,
                                `high_inf_pos`: Option[String] = None,
                                `motif_score_change`: Option[String] = None,
                                `impact`: String = "MODERATE",
                                `pick`: String = "1",
                                `variant_class`: String = "SNV",
                                `tsl`: Option[String] = None,
                                `hgvs_offset`: Option[String] = None,
                                `pheno`: Option[String] = None,
                                `minimised`: Option[String] = None,
                                `gene_pheno`: Option[String] = None,
                                `filter`: String = ".",
                                `flanking_bps`: String = "ACC",
                                `variant_id`: String = ".",
                                `variant_qual`: String = ".",
                                `tumortype`: String = "acyc",
                                `platform`: String = "exome",
                                `judgement`: String = "RETAIN",
                                `amino_acid_change`: String = "V334I",
                                `amino_acid_position`: String = "334",
                                `protein_lenght`: Option[String] = None,
                                `reference_amino_acid`: String = "V",
                                `variant_amino_acid`: String = "I",
                                `allele_freq`: String = "NA",
                                `tm`: String = "WARS2 334",
                                `amino_acid_length`: String = "360",
                                `ref_tri`: String = "ACC",
                                `oncotree_organtype`: String = "headandneck",
                                `oncotree_parent`: String = "saca",
                                `oncotree_detailed`: String = "acyc",
                                `master_id`: String = "000236")
