/**
 * Generated by [[bio.ferlab.datalake.spark3.utils.ClassGenerator]]
 * on 2022-04-06T14:34:50.380287
 */
package bio.ferlab.clin.model

import java.sql.Timestamp


case class EnrichedConsequences(`chromosome`: String = "1",
                                `start`: Long = 69897,
                                `reference`: String = "T",
                                `alternate`: String = "C",
                                `ensembl_transcript_id`: String = "ENST00000335137",
                                `ensembl_gene_id`: String = "ENSG00000186092",
                                `consequences`: List[String] = List("synonymous_variant"),
                                `vep_impact`: String = "LOW",
                                `symbol`: String = "OR4F5",
                                `ensembl_feature_id`: String = "ENST00000335137",
                                `feature_type`: String = "Transcript",
                                `strand`: Int = 1,
                                `biotype`: String = "protein_coding",
                                `exon`: EXON = EXON(),
                                `intron`: INTRON = INTRON(),
                                `hgvsc`: String = "ENST00000335137.4:c.807T>C",
                                `hgvsp`: String = "ENSP00000334393.3:p.Ser269=",
                                `cds_position`: String = "1000/1083",
                                `cdna_position`: String = "1027/2800",
                                `protein_position`: String = "334/360",
                                `amino_acids`: AMINO_ACIDS = AMINO_ACIDS(),
                                `codons`: CODONS = CODONS(),
                                `pick`: Boolean = true,
                                `original_canonical`: Boolean = true,
                                `refseq_mrna_id`: List[String] = List("NM_001005484.1", "NM_001005484.2"),
                                `aa_change`: String = "p.Ser269=",
                                `coding_dna_change`: String = "c.807T>C",
                                `impact_score`: Int = 2,
                                `created_on`: Timestamp = java.sql.Timestamp.valueOf("2022-04-06 13:41:31.039545"),
                                `updated_on`: Timestamp = java.sql.Timestamp.valueOf("2022-04-06 13:41:31.039545"),
                                `normalized_consequences_oid`: Timestamp = java.sql.Timestamp.valueOf("2022-04-06 13:41:31.039545"),
                                `consequence`: List[String] = List("synonymous"),
                                `predictions`: PREDICTIONS = PREDICTIONS(),
                                `conservations`: CONSERVATIONS = CONSERVATIONS(),
                                `consequences_oid`: Timestamp = java.sql.Timestamp.valueOf("2022-04-06 13:41:31.039545"),
                                `uniprot_id`: Option[String] = None,
                                `mane_select`: Boolean = false,
                                `mane_plus`: Boolean = false,
                                `canonical`: Boolean = false)

case class PREDICTIONS(`sift_converted_rank_score`: Option[Double] = None,
                       `sift_pred`: Option[String] = None,
                       `polyphen2_hvar_score`: Option[Double] = None,
                       `polyphen2_hvar_pred`: Option[String] = None,
                       `fathmm_converted_rankscore`: Option[Double] = None,
                       `fathmm_pred`: Option[String] = None,
                       `cadd_score`: Option[Double] = None,
                       `dann_score`: Option[Double] = None,
                       `revel_rankscore`: Option[Double] = None,
                       `lrt_converted_rankscore`: Option[Double] = None,
                       `lrt_pred`: Option[String] = None)

case class CONSERVATIONS(`phylo_p17way_primate_rankscore`: Option[Double] = None)
